### ISSUE TITLE
Context is not initialized

### ISSUE THREAD

**Issue by sdcdsvsdv423** [ryansolid/mobx-jsx#13]
I have component, which get context via `useContext` function:
```
 const [state, { setValue }] = useInput()
```
The problem is component factory invoking twice and first time I receive the empty state. Is it by design and I need to check whether context has already been activated?

**Comment by ryansolid** [comment]
Hmm.. factory invoking twice is the problem. The call to useContext won't do that so the problem must be higher up the tree either in the Component hosting the provider or the provider itself. We don't see this in the simple example I posted so I need to see more surrounding code to get an idea why it is running twice.

**Comment by sdcdsvsdv423** [comment]
I guess I realize what happens. I passed an observable object by properties, then some mutations in the parent component have affected to the child component factory.

But it's not pretty clear how to affect to the child component. For example, in Vue I can pass `props` and then call `$emit` for some events. Current approach is different and it's not so clear how to transfer data between parent and child components. Please can you clarify this point?

**Comment by ryansolid** [comment]
Parent to child uses props. But the props object in mobx-jsx is readonly. There is no 2-way binding. So while it is fine to bind to a property access as a prop to a component for reads, you can't write to it that way. It is the getter or setter that is reactive not the value. So in React MobX it is common to pass the parent object as the prop. https://mobx.js.org/best/pitfalls.html#dereference-values-as-late-as-possible. This is generally good advice but mobx-jsx is smart enough to automatically wrap prop expressions like that in a getter on the props object so you don't need to pass the parent unless you wish to write to it. But I think a better pattern would be to pass functions to the child that give them the ability to mutate the parents state.

Similarly, the other common pattern in React is to pass function handlers to children props as a means for the child to call back the parent. So, in the same way, you'd emit an event in Vue you'd make an onEvent handler. Ie...

```jsx
const ChildComponent = (props) => {
  return <div onClick={() => props.onCustom & props.onCustom()}>Child Content</div>
}

const ParentComponent = () => {
  return <ChildCompoment onCustom={console.log('custom event')} />
}
```
So basically everything is a prop going down. What you are describing should be fine on the surface. Passing observable properties down to the Child should not cause any weird retriggers of the whole Component. Basically everything is isolated. Any reactive expression (like property access on observable) that is in a JSX expression is wrapped in a getter on the props object and its access is deferred until that property is accessed. Consider the following:

```jsx
const UserDetails = () => {
  const state = observable({ firstName: "John", lastName; "Smith" });
  return (
    <>
      <NameTag displayName={state.firstName + " " + state.lastName} />
      <a onClick={() => state.firstName = "Jake"}>Change Name</a>
    </>
  )
}

const NameTag = (props) => <h2>{props.displayName}</h2>
```
It is worth pointing out NameTag basically compiles down to:
```js
const NameTag = (props) => {
  const el = document.createElement("h2");
  autorun(() => el.textContent = props.displayName);
  return el;
}
```

Now on initial render it's pretty obvious what the output should be. However, how it happens might not be obvious. It isn't until `displayName` is accessed off the props object(getter) inside `NameTag` that the expression assigned to displayName on the Component runs. In so the dependencies of the firstName and lastName access or actually tracked under the autorun inside NameTag. Nowhere else.
In so when you click on the "Change Name" anchor tag only that one autorun executes. Basically the text is replaced and nothing else reruns. In fact, at that point, the Component doesn't even exist anymore. If you look the only thing that exists after the function runs is the element and the autorun. That's why there are no lifecycles etc. Components are just an illusion over a graph of reactive and DOM nodes.

Anyway, hope that helps. This library doesn't have as good documentation but Babel Plugin JSX DOM Expressions as some. And Solid while has some specificity to Solid in terms of syntax basically works the same way. So most Docs examples basically apply here as well.

**Comment by esindger** [comment]
Please can you look at the [example](https://codesandbox.io/s/mobx-counter-context-0h6pd?fontsize=14)? Why dynamic component is not changing his state?

**Comment by sdcdsvsdv423** [comment]
I could reproduce this case. Please can you look at [example](https://codesandbox.io/s/mobx-counter-context-0h6pd?fontsize=14).

Context does not work if I return:
```js
return <>
  {DynamivComponent()}
</>;
```
State is not changing if I use the following sintax:
```js
return <>
  <DynamicComponent/>
</>;
```

**Comment by ryansolid** [comment]
The state not changing makes sense since Components are untracked context. To be fine grained we do not wrap Components in autorun just their bindings. They only run once. So that is the key difference.

However, Context is not working under a reactive context is the issue. I did some debugging and I see what's happening. MobX scheduling is messing things up. If already in an execution when it hits an `autorun` it doesn't run immediately like the docs say. However my library expects it to since it needs the initial output to diff subsequent value changes. I'm not sure if its possible to force immediate execution or if there is a way to change how things are written to pass this value another way. So going to look into it since this behavior has the potential to break some things. The Context mechanism for one. But even things like `lazy` or using Components to manage control flows.

On the positive other than the timing everything runs correctly. Negatively this doesn't seem easy to fix with passing around a getter function of some kind instead of a value which seems complex.

**Comment by ryansolid** [comment]
Well nvm I believe I fixed it with mobx-jsx v0.8.2. I believe I have fixed both nested reactivity and the issue with Context. Let me know if this helps with your issue here.

**Comment by sdcdsvsdv423** [comment]
Thank you!
